### ISSUE TITLE
Make RunCPM aware of and follow the $SYS and $R/O flags

### ISSUE THREAD

**Issue by MockbaTheBorg** [MockbaTheBorg/RunCPM#69]
Right now, if a file is renamed to an 8bit character on T1 and T2, STAT seems to see it as R/O and Hidden (SYS), but more coding is needed for it to work throughout the file system.

**Comment by MockbaTheBorg** [comment]
Closing .... this proved much more complicated than I thought. May require a full rewrite of RunCPM's filesystem. Giving up for now.

**Comment by mecparts** [comment]
At the risk of resurrecting a dead issue, I've just now got $SYS, $R/O, $ARC and $PUBLIC working in my ZCPR3.3 fork of RunCPM.

Background: months ago I modified the file system routines to examine the FAT file's attribute bits and set the corresponding bits in the FCB's filename and type. Hidden became Public (FN2), and ReadOnly, System and Archive slid into FT1, FT2, and FT3. With the SD card plugged into a Linux system, I could use fatattr to set the various bits, and when the card was plugged back into the RunCPM system, the altered BDOS would see (and respect) the attributes. That was enough to get NovaDOS's/Z80DOS's public file support working, as well as preventing writes to $R/O files and hiding $SYS files in directory listings.

But while there was support in SdFat for *reading* the file attributes, there wasn't any for *setting* them. Which was a bit of a pain. Any time I wanted to change an attribute I had to shut down the RunCPM system, pull the card, put it in the Linux box, fatattr the file, pull the card again, etc., etc. But tonight I sat down and starting going through the SdFat library with a fine tooth comb. I figured out that I could add a handful of routines alongside isHidden(), isReadOnly() and isSystem() to set the attributes: makeHidden(), makeReadOnly(), makeSystem() and makeArchive(). Each gets passed a boolean parameter indicating whether the attribute should be set or reset. In addition, each make...() routine also sets the file's dirty bit. So when the file is closed, and sync() is called, the directory entry gets updated. A few lines of code in sync() was enough to transfer the attribute bits from the FatFile instance into the directory entry and get them written to the SD card. So now STAT can change the $R/O and $SYS attributes, and PUBLIC can do the same to the $PUBLIC attributes.

Keep in mind, I've done this in my divergent ZCPR3.3 fork. It's fairly different than the master branch by now, *and* it's only been done for the Arduino side of things. It's *possible* that the same kind of mods I've made to get the attributes working could be done for the Windows build; after all, I've done this with FAT file attributes and that's an MSDOS thing. But I'm not at all sure how a Linux port would work. Assuming I don't discover that my "let's try this and see how it goes" mods to SdFat have ended up trashing the filesystem on my SD card, is it worthwhile thinking about backporting something that might end up only working on some systems?
